GUI Inventory Integration (Closes #41)

AnSiChen opened the pull request:
## PySeas: Inventory

This PR integrates an interactive GUI for managing the player's inventory, a **game's core feature**. It enhances user experience with features like item usage, removal, and real-time feedback while ensuring code quality through robust type-checking, formatting, and error handling, with added icons for further visual feedback.

[![pyseas-inventory.png](https://i.postimg.cc/G3ZXHdnK/pyseas-inventory.png)](https://postimg.cc/14rGdxtV)

### **Key Implementations**

**Inventory Management in GUI:**

1. Scrollable item display with "Use" and "Discard" buttons.
2. Real-time feedback for user actions displayed as messages.

**Testing:**

- Verified functionality with unit tests for inventory and GUI components.

### Progress on Issue #41 
**This PR addresses the following items:**

- [x]  The player can remove items from their inventory.
- [x]  The player can add items to their inventory.
- [x]  The player can open the inventory and view items.
- [x]  The player can use an item from their inventory, which removes it.

### Screenshots

**Icon Integration:**
[![icon-logic-implemented.png](https://i.postimg.cc/jdkyTsKs/icon-logic-implemented.png)

**Using an Item:** 
[![use-item.png](https://i.postimg.cc/9fK7JXpN/use-item.png)](https://postimg.cc/Y4N0SHT1)

**Removing an Item:** 
[![remove-item.png](https://i.postimg.cc/QCkK8QrR/remove-item.png)](https://postimg.cc/H8nk37n2)

### **Documentation**

**Inventory Guide: Testing and Structure**
1. Visual display of items with names and quantities.
2. Real-time interaction buttons and feedback.
3. Modify inventory.json to add items for testing.
4. Modify Initializer

**Utils Guide(Related Implementation):**
1. Explains messaging.py for JSON-based messages for Dynamic Messages.
2. Highlights modularity and expandability for future expansion or multilingual support.

### **Checklist**
 
- [x]  Features implemented as described.
- [x]  Code reviewed for type-checking and linting (mypy, ruff).
- [x]  Unit tests passed.
- [x]  Documentation updated with visuals and usage examples.

### Previously Also Considered:

**Implementing in Fewer Files**

- This approach was not pursued as it did not promote modularization or scalability.

**Using External Libraries for UI Design**

- While still being considered, this iteration focuses on addressing the issue using existing tools and frameworks.

**Using Placeholders for Icons**

- This PR implements the foundational logic to transition from placeholders to real Sprite Sheet Images.

### **Potential Future Enhancements**

- Expand inventory visuals (e.g., add "Buy item" and "Loot" features).
- Replace placeholder icons for certain inventory items to accurately represent each item in the Sprite Sheet.

ultimateownsz commented on docs/UtilsGuide.md:
I really like how this module could be expanded for multi-language support and integrated with other game systems like feedback or logging.

Here are a few thought I had while reading the documentation:
1. Your idea of replacing or extending data/messages.json with localized versions is a great idea. To make it even more contributor friendly, we could include examples how to structure localized JSON files (e.g, key-value pairs by language). Maybe we could add a placeholder for fallback languages, so the game always displays a message even if a translation is missing? There is one project I know that does this, maybe we could replicate this later down the line if we have more text in the game itself. 
2. I think integrating with game state management for dynamic message generation is a helpful feature to implement. 
3. I really like how you broke down the function parameters and explained the return value. The inclusion of an example `get_message(category: str, key: str, **kwargs) -> str` and the default error message adds a lot of clarity for others.

AnSiChen replied:
Great that you found it useful. I also believe this is a great feature, really scalable and easier to maintain.

ultimateownsz commented on src/GUI/gameloop.py:
Everything is well explained in the comments, nothing to add upon.

AnSiChen replied:
Thank you.

ultimateownsz commented on tests/test_inventory.py:
I had some problems at my end:

- ImportError: Failed to import test module: test_inventory
- from src.GUI.inventory import Inventory, Chest, Quest
ModuleNotFoundError: No module named 'src'
- Ran 1 test in 0.000s

FAILED (errors=1)

AnSiChen replied:
My apologies on this, totally my oversight, I ran the tests without issues, but after using Ruff and Mypy I add to add __init__ to several modules, and did not commit that part, I will check back the tests and add any missing files/code. Thank you very much, for all the feedback.

AnSiChen replied:
They were indeed extremely useful, I was thinking of incorporating them into my own projects, they helped me clean and format my code which increased their quality. The documentation was also good and fairly simple. It's just that I skipped some important steps. Good experience for when I attempt the fixes later!

ultimateownsz replied:
Thank you, yes the combination of mypy with ruff can be a time saver!

ultimateownsz reviewed (changes requested):
Hey, thank you for your pull request! I wrote down my thoughts on some things you need to adjust a bit. The unittest didn't work at my end. Perhaps I did something wrong. Let me know if so!

AnSiChen commented on data/inventory.json:
Hello, thank you very much for the feedback. You are right, Ruff and Mypy might have caused this, as all of them were well formatted, sorry about that, first time using both tools.

Concerning the inventory items and such, I went by the icons found in Treasure+.png, so as of now they are only placeholders, we would need to better define what we want in the inventory and better polish those details.  This was just to populate the inventory as we don't have a 'Buy' or similar in-game actions yet that would allow us to test the 'add_item'.

ultimateownsz replied:
> Hello, thank you very much for the feedback. You are right, Ruff and Mypy might have caused this, as all of them were well formatted, sorry about that, first time using both tools.

Hi, it could be some issues at how we set up ruff & mypy so for now it's fine. Need to look into it and fix it, perhaps, that JSON files are not being formatted like the rest of the codebase. Or have it differently formatted because it is a JSON file.

> Concerning the inventory items and such, I went by the icons found in Treasure+.png, so as of now they are only placeholders, we would need to better define what we want in the inventory and better polish those details. This was just to populate the inventory as we don't have a 'Buy' or similar in-game actions yet that would allow us to test the 'add_item'.

We have not properly discussed what items can be picked up, bought or stored in the inventory yet. For placeholder purposes it's fine, hadn't thought of this idea yet. Perhaps we can opt in our Discord community for helping us decide what items can be considered for use for the inventory and tackle this problem for in the future. Like for example having specific guidelines which item can be considered in the main game and have the community vote on it.

AnSiChen commented on docs/InventoryGuide.md:
Thank you gain. 

You are totally right, and we could totally explore SQLite of similar approaches, I was trying to focus on the basic functionalities first, to make sure we have a working feature, I also really like your suggestions in using something more solid and better integrated.

I'll get back to you concerning this via Discord.

ultimateownsz replied:
> Thank you gain.
> 
> You are totally right, and we could totally explore SQLite of similar approaches, I was trying to focus on the basic functionalities first, to make sure we have a working feature, I also really like your suggestions in using something more solid and better integrated.
> 
> I'll get back to you concerning this via Discord.

For this PR, you did already a great job integrating JSON with the inventory, I would not make changes to SQLite yet.

AnSiChen commented:
Hi all, I've pushed the latest changes based on feedback and updates.

### Summary of Updates:
- Added `__init__.py` files for package structure.
- Updated `tests/test_inventory.py` to fix `sys.path` for direct execution. (Alternatively use pytest)
- Revised `docs/InventoryGuide.md` to include test instructions.
- Added comments to `inventory_gui.py` to remind everyone of the nature of the current implementation.
  - (Once this last point has been addressed or we implement "Buy", "Add" or related in-game feature we can proceed to replace it. As explained in the PR discussion, we do not currently have a way to test the inventory, so for now it could be fine until we move forward.)

All tests, `ruff`, and `mypy` checks pass locally. Please let me know if further adjustments are needed. Thank you for reviewing!

ultimateownsz reviewed (commented):
Hey, thank you for your changes, before I can check your PR, there are some changes made to the codebase. You should use get pull from the main branch to see the latest changes. And update your own branch with those changes. Thanks in advance!

ultimateownsz reviewed (approved):
Thank you for your changes, sorry for the long wait, I was busy and was also sick for a few days. I will contact you in Discord for new tasks to pickup, my idea is to create a roadmap for PySeas to prioritize what needs to be worked on to create the working prototype! All input is appreciated!
